File: lib/utils/datetime_helper.dart
/// DateTime utilities for handling timezone conversions
/// Ensures all dates are in South African timezone (SAST - UTC+2)
library;

/// Get current DateTime in South African timezone (SAST - UTC+2)
/// 
/// South Africa uses SAST (South African Standard Time) which is UTC+2 year-round
/// (no daylight saving time)
DateTime getSouthAfricanTime() {
  final utcNow = DateTime.now().toUtc();
  // Add 2 hours for SAST (UTC+2)
  return utcNow.add(const Duration(hours: 2));
}

/// Convert any DateTime to South African timezone
DateTime toSouthAfricanTime(DateTime dateTime) {
  final utc = dateTime.toUtc();
  return utc.add(const Duration(hours: 2));
}

/// Get current date in South African timezone (without time)
DateTime getSouthAfricanDate() {
  final saTime = getSouthAfricanTime();
  return DateTime(saTime.year, saTime.month, saTime.day);
}

/// Check if a date is today in South African timezone
bool isTodaySA(DateTime date) {
  final today = getSouthAfricanDate();
  final checkDate = DateTime(date.year, date.month, date.day);
  return checkDate.isAtSameMomentAs(today);
}

/// Get start of today in South African timezone
DateTime getStartOfTodaySA() {
  final saDate = getSouthAfricanDate();
  return saDate;
}

/// Get end of today in South African timezone
DateTime getEndOfTodaySA() {
  final saDate = getSouthAfricanDate();
  return saDate.add(const Duration(days: 1)).subtract(const Duration(microseconds: 1));
}

File: mobile-rn/src/utils/datetimeHelper.ts
/**
 * Ported from lib/utils/datetime_helper.dart (verified against source on 2026-06-22).
 * South Africa uses SAST (UTC+2) year-round, no daylight saving.
 *
 * JS Date has no equivalent of Dart's `isUtc`-tagged-but-field-shifted DateTime trick the
 * source relies on. Instead, `getSouthAfricanTime()`/`toSouthAfricanTime()` return a Date
 * shifted by +2h whose *UTC* getters (`getUTCFullYear()`, `getUTCHours()`, etc.) read as
 * SAST wall-clock fields, regardless of the device's own local timezone — callers must
 * read fields via the UTC getters, not the local ones. `getSouthAfricanDate()` and the
 * start/end-of-day helpers instead return a real, correct absolute instant (safe to feed
 * straight into a Firestore Timestamp range query).
 */
const SAST_OFFSET_MS = 2 * 60 * 60 * 1000;

/** Mirrors getSouthAfricanTime(). Read fields via getUTC*() — see file-level comment. */
export function getSouthAfricanTime(): Date {
  return new Date(Date.now() + SAST_OFFSET_MS);
}

/** Mirrors toSouthAfricanTime(). Read fields via getUTC*() — see file-level comment. */
export function toSouthAfricanTime(date: Date): Date {
  return new Date(date.getTime() + SAST_OFFSET_MS);
}

/** Mirrors getSouthAfricanDate(): the real instant of SAST midnight today. */
export function getSouthAfricanDate(): Date {
  const sast = getSouthAfricanTime();
  return new Date(Date.UTC(sast.getUTCFullYear(), sast.getUTCMonth(), sast.getUTCDate()) - SAST_OFFSET_MS);
}

/** Mirrors isTodaySA(): whether `date`'s instant falls within today's SAST calendar day. */
export function isTodaySA(date: Date): boolean {
  const startOfToday = getSouthAfricanDate().getTime();
  const endOfToday = startOfToday + 86400000;
  return date.getTime() >= startOfToday && date.getTime() < endOfToday;
}

/** Mirrors getStartOfTodaySA(). */
export function getStartOfTodaySA(): Date {
  return getSouthAfricanDate();
}

/** Mirrors getEndOfTodaySA(). */
export function getEndOfTodaySA(): Date {
  return new Date(getSouthAfricanDate().getTime() + 86400000 - 1);
}
